peps/pep-0817.rst

```diff
@@ -0,0 +1,2528 @@
+endpoint to the existing package. In both cases, it may be necessary to
+preserve the old endpoint in minimal maintenance mode, to ensure that
+old wheels can still be installed. The old endpoint can trigger
+deprecation warnings in the ``get_all_configs()`` hook that is used when
+building packages.
+
+An alternative approach is to use semantic versioning to cut off
+breaking changes. However, this relies on package authors reliably using
+caps on dependencies, as otherwise old wheels will start using
+incompatible plugin versions. This is already a problem with Python
+build backends used today.
+
+When vendoring or reimplementing plugins, installers need to follow
+their current behavior. In particular, they should recognize the
+relevant provider versions numbers, and possibly fall back to installing
+the external plugin when the package in question is incompatible with
+the installer's implementation.
+
+
+Example use cases
+-----------------
+
+PyTorch CPU/GPU variants
+''''''''''''''''''''''''
+
+As of October 2025, `PyTorch
+<https://pytorch.org/get-started/locally/>`__ publishes builds a total
+of seven variants for every release: a CPU-only variant, three CUDA
+variants with different minimal CUDA runtime versions and supported
+GPUs, two ROCm variants and a Linux XPU variant.
+
+This setup could be improved using GPU/XPU plugins that query the
+installed runtime version and installed GPUs/XPUs to filter out the
+wheels for which the runtime is unavailable, it is too old or the user's
+GPU is not supported, and order the remaining variants by the runtime
+version. The CPU-only version is published as a null variant that is
+always supported.
+
+If a GPU runtime is available and supported, the installer automatically
+chooses the wheel for the newest runtime supported. Otherwise, it falls
+back to the CPU-only variant. In the corner case when multiple
+accelerators are available and supported, PyTorch package maintainers
+indicate which one takes preference by default.
+
+
+Optimized CPU variants
+''''''''''''''''''''''
+
+Wheel variants can be used to provide variants requiring specific CPU
+extensions, beyond what platform tags currently provide. They can be
+particularly helpful when runtime dispatching is impractical, when the
+package relies on prebuilt components that use instructions above the
+baseline, when availability of instruction sets implies library ABI
+changes, or simply to benefit from compiler optimizations such as
+auto-vectorization applied across the code base.
+
+For example, an x86-64 CPU plugin can detect the capabilities for the
+installed CPU, mapping them onto the appropriate x86-64 architecture
+level and a set of extended instruction sets. Variant wheels indicate
+which level and/or instruction sets are required. The installer filters
+out variants that do not meet the requirements and select the best
+optimized variant. A non-variant wheel can be used to represent the
+architecture baseline, if supported.
```

Comment:
Earlier the suggestion was to use the `null` variant. Is this different advice? If not, could it be clarified?